### DISCUSSION TITLE
Creating an application installer library

### DISCUSSION THREAD

**Discussion by aaronparker** [EUCPilots/evergreen-module#357] [Ideas]
I’m adding some functionality to Evergreen that it would be good to get some feedback on - is this useful, what else would you like to see this functionality provide?

With Evergreen you could build a library of app installers for image builds, so that you can install the latest version or rollback to a previous version. So far it looks like this:
* `Export-EvergreenApp` - exports the application version information from `Get-EvergreenApp` to a JSON file. This function automatically updates an existing file while removing duplicates 
* `New-EvergreenLibrary` - creates a directory and manifest file where you can build a library of installers. The manifest file stores information about the applications to query via Evergreen
* `Invoke-EvergreenLibraryUpdate` - queries the library manifest, finds the application version info, downloads the installers and then saves the application version info to the JSON file for future reference. This function could be run via a scheduled task to automatically update the library

Here’s what a library might look like:

```powershell
E:\EvergreenLibrary\EvergreenLibrary.json
E:\EvergreenLibrary\MicrosoftEdge\MicrosoftEdge.json
E:\EvergreenLibrary\MicrosoftEdge\Stable\Enterprise\103.0.1264.62\x64\MicrosoftEdgeEnterpriseX64.msi
E:\EvergreenLibrary\MicrosoftOneDrive\MicrosoftOneDrive.json
E:\EvergreenLibrary\MicrosoftOneDrive\Production\22.131.0619.0001\AMD64\OneDriveSetup.exe
E:\EvergreenLibrary\MicrosoftOneDrive\Production\22.141.0703.0002\AMD64\OneDriveSetup.exe
```

The library manifest file lists the applications in the library. For each application we have the name, which is the application name supported by Evergreen and the top-level directory for that app, as well as the filter used to define the application installers to include in the library:

```json
[
    {
        "Name": "EvergreenLibrary",
        "Applications": [
            {
                "Name": "MicrosoftOneDrive",
                "Filter": "$_.Architecture -eq \"AMD64\" -and $_.Ring -eq \"Production\""
            },
            {
                "Name": "MicrosoftEdge",
                "Filter": "$_.Platform -eq \"Windows\" -and $_.Channel -eq \"Stable\" -and $_.Release -eq \"Enterprise\" -and $_.Architecture -eq \"x64\""
            }
        ]
    }
]
```

You might create multiple libraries - one library would store production installers, while another library might host preview installers for testing.

**Comment by JimMoyle** [comment]
Are you thinking of using this as part of a private Winget repository?

**Reply by aaronparker** [reply]
I assume that could be a use-case, but I've not looked at the details. Could also be for being able to install an older version of an app, separating app downloads form the image pipeline build, or for an application packaging factory
